### PR TITLE
Eliminate undefined behaviour for amd64 builds

### PR DESCRIPTION
**Pull request by lonkamikaze**
I used clang to build with canaries for undefined behaviour and eliminated all instances triggered, with opengl1, opengl1, vm_*=0, vm_*=1 and vm_*=2.

Note e52d30e, this has apparently never worked as intended, so it actually fixes the functionality and the undefined behaviour.

### CODE REVIEW

File: code/jpeg-8c/jidctint.c [hunk 1]
```diff
@@ -206,7 +206,7 @@ jpeg_idct_islow (j_decompress_ptr cinfo, jpeg_component_info * compptr,
 	inptr[DCTSIZE*5] == 0 && inptr[DCTSIZE*6] == 0 &&
 	inptr[DCTSIZE*7] == 0) {
       /* AC terms all zero */
-      int dcval = DEQUANTIZE(inptr[DCTSIZE*0], quantptr[DCTSIZE*0]) << PASS1_BITS;
+      int dcval = (unsigned int)DEQUANTIZE(inptr[DCTSIZE*0], quantptr[DCTSIZE*0]) << PASS1_BITS;
```
Comment (timangus) [inline code/jpeg-8c/jidctint.c]: Suspicious looking cast?

Comment (lonkamikaze) [inline code/jpeg-8c/jidctint.c]: Not at all. This is necessary to make `<< PASS1_BITS` legal.

**Comment by smcv** [comment]
Have you sent the libjpeg8 UB fixes upstream to IJG? That way all libjpeg users would benefit (if there are many left, now that Linux distributions are basically all using the libjpeg-turbo fork).

Ideally the bundled third-party libraries in ioquake3 (other than minizip, which specifically has local modifications) would be verbatim copies of their upstream counterparts. (... Or not exist, but I recognise that that would be really awkward for building on Windows.)

**Comment by lonkamikaze** [comment]
@smcv No I haven't, AFAIK libjpeg is dead.

**Comment by smcv** [comment]
> AFAIK libjpeg is dead. Is there an active project, somewhere?

I'd characterize it as semi-dead. Next time ioquake3 updates its bundled library I'd suggest libjpeg-turbo, which is what's in Firefox and all the major Linux distributions.

The websites for the original libjpeg (from which libjpeg-turbo forked) seem to be variously http://jpegclub.org/ http://www.ijg.org/ and http://www.infai.org/jpeg/ (but as far as I know it's just one developer claiming to speak for a larger group, and occasionally adding non-interoperable extensions to the format).

**Comment by zturtleman** [comment]
I cherry picked [most of the commits](https://github.com/ioquake/ioq3/compare/9f294ce5207fa29b30d6b004665e44be3af3dc8e...b8ee77cecc749139c04d7d3684ff5d1f0972a53b) that are still relevant. Some of the code was fixed or removed since this PR was opened. I didn't merge the two libjpeg commits or commit for OP_MULI in vm_interpeted.c. OP_MULI change would presumably could cause different behavior between native libs and QVMs. Thanks for submitting the pull request.

559193ce2b128901f3eac310bc5859e1af10b747 Shifting negative integers is undefined behaviour
45e6d6386978c5a822f5abdb6ae9a6570c92ce48 Stop shifting negative integers
4f72b81d51b39c7a7d40122c8b816ccf265ae36a Avoid signed integer overflow

**Pull request closed without merging**